timemachine/md/smc.py

```diff
@@ -19,144 +19,190 @@
 IndexArray = Array
 LogWeights = Array
 First = bool
+Iteration = int
 
 BatchPropagator = Callable[[Samples, Lambda], Samples]
-BatchLogProb = Callable[[Samples, Lambda], LogWeights]
-BatchLogProbASMC = Callable[[Samples, Lambda, First], LogWeights]
+BatchLogProb = Callable[[Samples, Lambda, First], LogWeights]
+FindNextLambda = Callable[[Samples, Lambda, Iteration, LogWeights], tuple[Lambda, LogWeights]]
 
 Resampler = Callable[[LogWeights], tuple[IndexArray, LogWeights]]
 ResultDict = dict[str, Any]
 
 
+class SMCMaxIterError(Exception):
+    """
+    Exception when SMC exceeds the maximum number of iters.
+    """
+
+    pass
+
+
 def sequential_monte_carlo(
     samples: Samples,
-    lambdas: Array,
     propagate: BatchPropagator,
     log_prob: BatchLogProb,
     resample: Resampler,
+    find_next_lambda: FindNextLambda,
+    store_intermediate_traj=True,
 ) -> ResultDict:
-    """barebones implementation of Sequential Monte Carlo (SMC)
+    """Implementation of Adaptive Sequential Monte Carlo (SMC).
+       This will adaptively interpolate between lambda=0 and lambda=1,
+       starting at lambda=0.
```

Comment:
May want to revise the docstrings of `sequential_monte_carlo` or `adaptive_find_next_lambda` to avoid duplication.